server/container.go

```diff
@@ -411,10 +411,50 @@ func (s *Server) RemoveContainer(ctx context.Context, req *pb.RemoveContainerReq
 	return &pb.RemoveContainerResponse{}, nil
 }
 
+// filterContainer returns whether passed container matches filtering criteria
+func filterContainer(c *pb.Container, filter *pb.ContainerFilter) bool {
+	if filter != nil {
+		if filter.State != nil {
+			if *c.State != *filter.State {
+				return false
+			}
+		}
+		// TODO(mrunalp): Add support for label filtering
+	}
+	return true
+}
+
 // ListContainers lists all containers by filters.
 func (s *Server) ListContainers(ctx context.Context, req *pb.ListContainersRequest) (*pb.ListContainersResponse, error) {
 	var ctrs []*pb.Container
-	for _, ctr := range s.state.containers.List() {
+	filter := req.Filter
+	ctrList := s.state.containers.List()
+
+	// Filter using container id and pod id first.
+	if filter != nil {
```

Comment:
we would likely do this sort of stuff with reducer funcs (just thinking out loud)